src/main/java/org/wso2/carbon/hash/provider/pbkdf2/PBKDF2HashProvider.java

```diff
@@ -0,0 +1,213 @@
+            }
+            throw new HashProviderServerException(ErrorMessage.ERROR_CODE_INVALID_KEY_SPEC.getDescription(),
+                    Constants.IDENTITY_HASH_PROVIDER_PBKDF2_ERROR_PREFIX +
+                            ErrorMessage.ERROR_CODE_INVALID_KEY_SPEC.getCode());
+        }
+    }
+
+    /**
+     * This method is responsible fpr validating the value to be hashed.
+     *
+     * @param value The value which needs to be hashed.
+     * @throws HashProviderClientException If the hash value is not provided.
+     */
+    private void validateEmptyValue(String value) throws HashProviderClientException {
+
+        if (StringUtils.isBlank(value)) {
+            throw new HashProviderClientException(
+                    ErrorMessage.ERROR_CODE_EMPTY_VALUE.getDescription(),
+                    Constants.IDENTITY_HASH_PROVIDER_PBKDF2_ERROR_PREFIX +
+                            ErrorMessage.ERROR_CODE_EMPTY_VALUE.getCode());
+        }
+    }
+
+    /**
+     * This method is responsible for validating the salt.
+     *
+     * @param salt The salt which needs to be validated.
+     * @throws HashProviderClientException If the salt value is blank.
+     */
+    private void validateSalt(String salt) throws HashProviderClientException {
+
+        if (StringUtils.isBlank(salt)) {
+            throw new HashProviderClientException(
+                    ErrorMessage.ERROR_CODE_SALT_EMPTY_EXCEPTION.getDescription(),
+                    Constants.IDENTITY_HASH_PROVIDER_PBKDF2_ERROR_PREFIX +
+                            ErrorMessage.ERROR_CODE_EMPTY_VALUE.getCode());
+        }
+    }
+
+    /**
+     * Resolve the iteration count according to the given meta properties.
+     *
+     * @param properties The properties map.
+     * @return The iteration count.
+     */
+    private int resolveIterationCount(Map<String, Object> properties) {
+
+        if (!properties.containsKey(Constants.ITERATION_COUNT_PROPERTY)) {
+            return Constants.DEFAULT_ITERATION_COUNT;
+        }
+        return (int) properties.get(Constants.ITERATION_COUNT_PROPERTY);
+    }
+
+    /**
+     * Resolve derived key length from the metaProperties.
+     *
+     * @param properties The properties map.
+     * @return The derived key length.
+     */
+    private int resolveDerivedKeyLength(Map<String, Object> properties) {
+
+        if (!properties.containsKey(Constants.DERIVED_KEY_LENGTH_PROPERTY)) {
+            return Constants.DEFAULT_DERIVED_KEY_LENGTH;
+        }
+        return (int) properties.get(Constants.DERIVED_KEY_LENGTH_PROPERTY);
+    }
+
+    /**
+     * Resolve pseudo random function from the metaProperties.
+     *
+     * @param properties The properties map.
+     * @return The pseudo random function.
+     */
+    private String resolvePseudoRandomFunction(Map<String, Object> properties) {
+
+        if (!properties.containsKey(Constants.PSEUDO_RANDOM_FUNCTION_PROPERTY)) {
+            return Constants.DEFAULT_PBKDF2_PRF;
```

Comment:
When we use this method when calculating the hash, we need to get the initialized value, not the default value.

Reply:
Fix other methods as well

Reply:
Even though in the initialized value, there is a possibility to have `dkLength` and `iterationCount` but not `pseudoRandomFunction`. In that case, it will set the `pseudoRandomFunction` as the default value.

Reply:
The `properties` map will be the map that is used to initialize the params. So if the `pseudoRandomFunction` was not initialized, then it will return the default value.
```
public void init(Map<String, Object> initProperties) {

        iterationCount = resolveIterationCount(initProperties);
        dkLength = resolveDerivedKeyLength(initProperties);
        pseudoRandomFunction = resolvePseudoRandomFunction(initProperties);
    }
```

Reply:
validated params `iterationCount`, `dkLength` and `pseudoRandomFunction` inside the `init` method itself and `resolveIterationCount` method and other methods also altered such that to return initialized parameters when the particular param is null in the metaProperties map.